[4.1] Ravendb 8113

aviv86 opened the pull request.

ayende commented on src/Raven.Client/Documents/Operations/Counters/CounterBatch.cs:
```diff
@@ -0,0 +1,28 @@
+        public List<CounterOperation> Counters = new List<CounterOperation>();
+    }
+
+    public class CounterOperation 
```
This is _confusing_ `CounterOperation` vs `CountersOperation`.
Also `GetOrDeleteCounters` ? 

Let's do it: `CounterOperation  that has a `Mode` (`Set`, `Increment`, `Delete`, `Get`)

ayende commented on src/Raven.Client/Documents/Operations/Counters/CountersDetail.cs:
```diff
@@ -0,0 +1,49 @@
+
+        public CounterDetail()
+        {
+            CounterValues = new Dictionary<string, long>();
```
`CounterValues` should default to `null`. 
There is a distinction between empty and not there.

ayende commented on src/Raven.Client/Documents/Operations/Counters/DeleteCounterOperation.cs:
```diff
@@ -19,32 +22,49 @@ public DeleteCounterOperation(string documentId, string name)
+            private readonly DocumentConventions _conventions;
 
-            public DeleteCounterCommand(string documentId, string name)
+            public DeleteCounterCommand(string documentId, string name, DocumentConventions conventions)
```
I think we can just move it all to the batch.
We'll probably still want to have simple incr support with POST and query string parameters.
To make it easy to hack it together, though

ayende commented on src/Raven.Client/Documents/Operations/CountersOperationExecutor.cs:
```diff
@@ -0,0 +1,64 @@
+        {
+            _store = store;
+            _databaseName = databaseName ?? store.Database;
+            _operations = _store.Operations.ForDatabase(_databaseName);
```
How would you control load balancing here?

ayende commented on src/Raven.Server/Documents/Handlers/CountersHandler.cs:
```diff
@@ -24,75 +26,98 @@ public class CountersHandler : DatabaseRequestHandler
+
             public long CurrentValue;
-            public bool DocumentMissing;
+            public (bool missing, string missingId, string counterName) DocumentMissing;
```
We use `PascalCasing` for tuple names

ayende commented on src/Raven.Server/Documents/Handlers/CountersHandler.cs:
```diff
@@ -24,75 +26,98 @@ public class CountersHandler : DatabaseRequestHandler
-                        var data = context.ReadObject(doc.Data, _doc, BlittableJsonDocumentBuilder.UsageMode.ToDisk);
-                        _database.DocumentsStorage.Put(context, _doc, null, data, flags: DocumentFlags.HasCounters);
+                        value += val;
+                        if (fullValues != null)
```
Why are you setting this?

ayende commented on src/Raven.Server/Documents/Handlers/CountersHandler.cs:
```diff
@@ -24,75 +26,98 @@ public class CountersHandler : DatabaseRequestHandler
+                        DocumentId = counter.DocumentId,
+                        CounterName = counter.CounterName,
+                        TotalValue = value,
+                        CounterValues = fullValues
```
Do we actually need all the values here? Can we avoid this except for the studio?

ayende commented on src/Raven.Server/Documents/Handlers/CountersHandler.cs:
```diff
@@ -190,184 +225,97 @@ private void UpdateDocumentCounters(BlittableJsonReaderObject metadata)
         }
 
-        [RavenAction("/databases/*/counters", "DELETE", AuthorizationStatus.ValidUser)]
+        [RavenAction("/databases/*/counters/delete", "POST", AuthorizationStatus.ValidUser)]
```
Don't do `/delete` in URLs.
We'll be moving it all to the batch soon anyway

ayende commented on test/SlowTests/Client/Counters/CountersCrudSingleNode.cs:
```diff
@@ -19,64 +19,91 @@ public void IncrementCounter()
                     session.SaveChanges();
                 }
+                var a = store.Counters.IncrementAsync("users/1-A", "likes", 5);
+                var b = store.Counters.IncrementAsync("users/1-A", "likes", 10);
```
This looks great

ayende commented on test/SlowTests/Client/Counters/CountersCrudSingleNode.cs:
```diff
@@ -19,64 +19,91 @@ public void IncrementCounter()
-                var a = store.Operations.SendAsync(new IncrementCounterOperation("users/1-A", "likes", 5));
-                var b = store.Operations.SendAsync(new IncrementCounterOperation("users/1-A", "likes", 10));
-                Task.WaitAll(a, b); // run them in parallel and see that they are good
+                var countersDetail = store.Operations.Send(new GetCounterValuesOperation(new GetOrDeleteCounters
```
Need nice API for this as well

aviv86 committed:
RavenDB-8113 : fix PR comments

aviv86 committed:
RavenDB-8113 Counters : refactor server side

aviv86 committed:
RavenDB-8113 Counters : refactor client api

aviv86 committed:
RavenDB-8113 : remove GetCountersForDoc test, add test case for MultiGet

aviv86 committed:
RavenDB-8113 : fix bug in IncrementCounterCommand.UpdateDocumentCounters()

aviv86 committed:
RavenDB-8113 : CountersHandler and CountersBatch refactoring

aviv86 committed:
RavenDB-8113 : nicer client api

ayende commented on src/Raven.Client/Documents/Operations/CountersOperationExecutor.cs:
```diff
@@ -0,0 +1,131 @@
+
+        public async Task<CountersDetail> BatchAsync(CounterBatch counterBatch)
+        {
+            return await _operations.SendAsync(new CounterBatchOperation(counterBatch)).ConfigureAwait(false);
```
Can skip `await` here

ayende commented on src/Raven.Client/Documents/Operations/CountersOperationExecutor.cs:
```diff
@@ -0,0 +1,131 @@
+                }
+            };
+
+            return await _operations.SendAsync(new CounterBatchOperation(counterBatch)).ConfigureAwait(false);
```
Can skip `await` here

ayende commented on src/Raven.Client/Documents/Operations/CountersOperationExecutor.cs:
```diff
@@ -0,0 +1,131 @@
+                }
+            };
+
+            await _operations.SendAsync(new CounterBatchOperation(counterBatch)).ConfigureAwait(false);
```
Can skip `await` here

aviv86 committed:
RavenDB-8113 CountersHandler :
- remove @counters property from metadata if doc no longer has counters
- send HasCounters flag to PUT only if document still have counters..
